MIDI-99/increase dataset size

WojciechMat opened the pull request:
### Code refactor, optimization etc.
 **Use of large datasets**:
1. Sharding dataset to parts with less rows than 5000. They are processed seperately, saved to disc and than concatenated to one big dataset.
2.  `dataset_name` is now stored inside `dataset.info.description` to append it to sequences' sources in tokenized dataset. This allows `dataset.shard` method to work properly ([here](https://github.com/Nospoko/midi-translation/pull/11/files#diff-74cad45aefddbfaa7bb449f5cacab993634353f0add41f56dcee2ce9d185cd42R141)).
3. Different method of slicing quantized piece to records in [`quantized_piece_to_records`](https://github.com/Nospoko/midi-translation/pull/11/files#diff-74cad45aefddbfaa7bb449f5cacab993634353f0add41f56dcee2ce9d185cd42R51) - one that does not include creating a MidiPiece object, takes ~2 times less time on giant-midi-sustain
4. You can train on multiple datasets, test+validation of maestro-v1-sustain is always used as a val_dataset.

E.g.
```shell
python train.py train.warmup=4000 dataset_name=roszcz/giant-midi-sustain+roszcz/maestro-v1-sustain
```

**Other small refactors**: 
1. Warmup hyperparameter was moved to "train" field, because I think it is where it should have been as it is used in lr_scheduler. 
3. Translation datasets are created in train.py instead of inside the main functions of the dstart and velocity pipeline to avoid repetition.

### Augmentation

Pitch shift and speed change are the augmentation methods you can use.
Augmenting is done using all cpus.

You can specify `augmentation.probability`, `augmentation.repetitions` (number of trials of augmenting individual record from translation dataset) and `augmentation.shift` (maximum pitch shift)
e.g.
```shell
python train.py augmentation.probability=0.5 augmentation.repetitions=2 augmentation.shift=3
```

Code snippet for loading a train dataset and augmenting it:
```python

datasets = []
for name in cfg.dataset_name.split("+"):
    dataset = load_cache_dataset(
        dataset_cfg=cfg.dataset,
        dataset_name=name,
        split="train",
    )
    datasets.append(dataset)

train_dataset = concatenate_datasets(datasets)

if cfg.augmentation.repetitions > 0:
    train_translation_dataset = augment_dataset(
        dataset=train_translation_dataset,
        dataset_cfg=cfg.dataset,
        augmentation_cfg=cfg.augmentation,
    )

train_dataset = MyTokenizedMidiDataset(
    dataset=train_translation_dataset,
    dataset_cfg=cfg.dataset,
    src_encoder=src_encoder,
    tgt_encoder=tgt_encoder,
)
```

WojciechMat committed:
dstart prediction

WojciechMat committed:
cosmetic changes

WojciechMat committed:
bin edges

WojciechMat committed:
make training possible for giant-midi-sustain

WojciechMat committed:
augmentation

WojciechMat committed:
augmentation

WojciechMat committed:
optimize quantized_piece_to_records

WojciechMat commented:
I decided to only use built-in data classes and random package in augmentation.py. 
In current approach I want to cache augmented datasets - now I'm thinking that maybe I shoud be able to augment dataset after loading it from the cache.
I left time measurement in the last commit - will remove it later.
Augmentation time for maestro-v1-sustain is ~0.53 seconds.
With probability=0.5 it produces ~1700 pieces from original 962 on train split when trying to augment every record 1 time.

I also optimized quantized_piece_to_records:
By using piece[start:end] it created new MidiPiece obj each time I think. 
I use only the existing dataframe to run through the piece now, works ~2 times faster (did not measure)

WojciechMat committed:
augmentation right before MyTokenizedMidiDataset

WojciechMat committed:
augmentation right before MyTokenizedMidiDataset

WojciechMat commented:
I moved augmentation to after loading a cached dataset.
Usage:
```py
dataset = load_cached_dataset(dataset_cfg, dataset_name, split)
augmented_dataset = augment_dataset(
        dataset=dataset,
        dataset_cfg=dataset_cfg,
        augmentation_probability=0.5,
        augmentation_rep=1,
    )
tokenized_dataset = MyTokenizedMidiDataset(
        dataset=augmented_dataset,
        dataset_cfg=dataset_cfg,
        src_encoder=src_encoder,
        tgt_encoder=tgt_encoder,
    )
```
For example: on train split of maestro-v1-sustain with augmentation probability of 0.5 and augmentation repetitions (how many times does the program try to augment each record) of 1, it produced 231043 records compared to previous 132318 in original dataset. Augmentation took 1.5 minutes. 

Asking for general review before moving on to the next steps.

WojciechMat commented:
Rebased this branch to master.

roszcz commented on data/dataset.py:
```diff
@@ -48,23 +48,24 @@ def build_translation_dataset(
         start = jt * sequence_step
         finish = start + sequence_len
-        part = piece[start:finish]
+        part = df.iloc[start:finish]
```
Using `MidiPiece` here allows us to track the source exactly, by switching to `df` we make the information about start and finish lost and impossible to track. 

Let me know if this change was to improve performance, if not it would be better to use `piece` here, if yes you can manually add start and finish to the source.

roszcz commented on pipeline/dstart/main.py:
```diff
@@ -1,22 +1,42 @@
         dataset_cfg=cfg.dataset,
-        dataset_name=cfg.dataset_name,
-        split="validation",
+        augmentation_probability=0.5,
+        augmentation_rep=1,
```
We should be able to control those values, rather than hardcoding them :)

WojciechMat committed:
add start and finish to source

WojciechMat commented:
Previos slicing quantized pieces took 3:06 minutes on maestro-v1-sustain, now it takes 1:13 minutes. I append start and finish manually to the source.

WojciechMat committed:
no val data augmentation, only augment when cfg.augmentation_reps>0

WojciechMat commented:
moved augmentation_rep and augmentation_probability to configs.

I deleted validation dataset augmentation for now - I think it will be better to use the same validation dataset for each model - we will be able to compare val loss across augmented dataset sizes during training.

WojciechMat committed:
do not cache dataset option

WojciechMat commented:
I have added option to not save a dataset - it kills process on kaggle when using giant-midi-sustain (and possibly other large datasets). I think we will need to adress that issue later as it also kills the process on my computer :thinking: 
perhabs it is the same issue as [this](https://discuss.huggingface.co/t/working-with-large-datasets/1876)
Edit: process is still getting killed, trying to find the issue now.

roszcz commented:
@WojciechMat I had some issues with too large datasets as well, and what I ended doing was something like this:

```python
for it in range(n_parts):
    part_of_data = dataset.get_small_part(it)
    new_hf_dataset = process(part_of_data)
    new_hf_dataset.save_to_disk("tmp/new_dataset_part_{it}")

# Later, maybe in different process
huge_dataset = concatenate_datset([Dataset.from_disk("tmp/new_dataset_part_{it}") for it in range(n_parts)])
hug_dataset.save_to_disk("tmp/new_dataset")
```

If your process is getting killed, it's probably Out Of Memory situation - HF should offer us the ability to operate datasets from disk, but I was not able to figure out the best approach to do this 🤔

WojciechMat committed:
sharding dataset before building

WojciechMat commented:
I harcoded 5000 as maximum shard size, bigger datasets are splitted to shards, shard directories are deleted after concatenating to a huge dataset to save disc space (also, they are not needed after).

WojciechMat committed:
fix deleting shard files

WojciechMat committed:
fix :)

WojciechMat committed:
make removing directories work for real

WojciechMat committed:
do not delete shard files ig?

WojciechMat commented:
it turned out that deleting shard directories caused problems with Dataset.train_test_split :man_shrugging: 
will think about it.

roszcz commented on data/augmentation.py:
```diff
@@ -0,0 +1,99 @@
+        done = True
+
+    # if no augmentation was done, return None
+    return augmented if done else None
```
I think we could have separate parameters for pitch shift probability and tempo change probability. Then you could have two separate methods doing one well defined thing each, and you won't need this akward "done" flag :)

roszcz commented on data/augmentation.py:
```diff
@@ -0,0 +1,99 @@
+    # shift pitch augmentation
+    if random.random() < augmentation_probability:
+        # max shift is octave down or up
+        shift = random.randint(1, 12)
+        augmented["pitch"] = pitch_shift(augmented["pitch"], shift)
```
Looks like you maybe misunderstood what `pitch_shift` is doing, and what the role `shift_threshold` is (it would be easier to spot if you used named parameters ;))

`shift_threshold` is controlling the maximum random pitch shift, there's no point in setting it randomly, as you could just set it to 12 here and get the same behaviour. 

The idea is that pitch shifting music too much is doing to much damage to the musical content (i.e. chopin shifted 2 octaves would be ridiculous), and we should treat it as a hyper parameter that we'll try to figure out based on the results we see.

roszcz commented on pipeline/velocity/main.py:
```diff
@@ -9,16 +10,35 @@ def load_datasets(cfg: DictConfig) -> tuple[MyTokenizedMidiDataset, MyTokenizedM
-        dataset_name=cfg.dataset_name,
-        split="validation",
-    )
+    if "maestro" in cfg.dataset_name:
```
We shouldn't do ifs like that - it's better to find a way to opearate on split-agnostic dataset objects.

roszcz reviewed (commented):
Thanks, I left some comments :)

As a general note, can you always keep the PR description up to date, and act as a small tutorial on how to execute/run/test the changes that the PR is introducing? It will make it easier to review and test locally :)

roszcz commented on data/dataset.py:
```diff
@@ -117,15 +122,40 @@ def add_cls_token(self, src_token_ids: list[int], tgt_token_ids: list[int]):
+    num_shards: int = 2,
+) -> Dataset:
+    shard_paths = []
+    for shard_i in range(num_shards):
+        path = f"{dataset_cache_path}-part-{shard_i}"
+        dataset_shard = dataset.shard(num_shards=num_shards, index=shard_i)
+        print(f"Processing shard {shard_i + 1} of {num_shards} with {len(dataset_shard)} records.")
```
```suggestion
    for it in range(num_shards):
        path = f"{dataset_cache_path}-part-{it}"
        dataset_shard = dataset.shard(num_shards=num_shards, index=it)
        print(f"Processing shard {it} of {num_shards} with {len(dataset_shard)} records.")
```

That's my personal preference for naming for loop range iterators :+1:, and we shouldn't switch to 1-based indexing just for prints

roszcz commented on data/dataset.py:
```diff
@@ -144,10 +174,18 @@ def load_cache_dataset(
+    # hardcoded maximum shard size as 5000
+    num_shards = len(midi_dataset) // 5000 + 1
+    translation_dataset = shard_and_build(
         dataset=midi_dataset,
+        dataset_name=dataset_name,
         dataset_cfg=dataset_cfg,
+        num_shards=num_shards,
+        dataset_cache_path=dataset_cache_path,
     )
-    translation_dataset.save_to_disk(dataset_cache_path)
+
```
This has  too many arguments. You don't need to pass `dataset_name` (if you need to attach this information after building, you can do it outside of `shard_and_build`) and if you're hardocing sharding parameters, it's better to hide it inside the sharding function.

WojciechMat replied:
We need to add dataset_name to dataset.info before build_translation_dataset, because we append dataset.info.dataset_name to each sequence.

roszcz replied:
Are you sure you need to add it though? I tried locally, and looks like shards are inheriting the parent dataset info. But if that's not the case, I think you still don't need to do this inside `shard_and_build`, but rather after it:

```python
translation_dataset = shard_and_build(...)
translation_dataset.info.dataset_name = dataset_name
```

That way you'll avoid unnecessary obfuscation

WojciechMat replied:
I had some problems with sharding when I added it before, I will try again, thanks!

WojciechMat replied:
Adding dataset_name to dataset.info before sharding:
```python
midi_dataset.info.dataset_name = dataset_name
    # hardcoded maximum shard size as 5000
    num_shards = len(midi_dataset) // 5000 + 1
    translation_dataset = shard_and_build(
        dataset=midi_dataset,
        dataset_name=dataset_name,
        dataset_cfg=dataset_cfg,
        num_shards=num_shards,
        dataset_cache_path=dataset_cache_path,
    )
```
produces an error when calling dataset.shard:
TypeError: DatasetInfo.__init__() got an unexpected keyword argument 'dataset_name'

and btw I am lost as to why did it work before: maestro-v1-sustain also does not seem to have this attribute inside its "info".

roszcz commented on data/augmentation.py:
```diff
@@ -0,0 +1,99 @@
+        n_velocity_bins=dataset_cfg.quantization.velocity,
+    )
+    all_records = []
+    for it, record in tqdm(enumerate(dataset), total=len(dataset)):
+        augmented_records = [
+            apply_augmentation(
+                record=record,
+                quantizer=quantizer,
+                augmentation_probability=augmentation_probability,
+            )
+            for _ in range(augmentation_rep)
+        ]
+        records = [aug for aug in augmented_records if aug is not None] + [record]
+        all_records += records
```
Can you look into running this on multiple cpus? I think HF Dataset provides some helper functions for that, and we should be able to run this ~10x faster

roszcz commented:
> it turned out that deleting shard directories caused problems with Dataset.train_test_split 🤷‍♂️ will think about it.

Can you share a code snippet demonstrating the issue?

WojciechMat commented:
> Can you share a code snippet demonstrating the issue?

The error comes from this line:
https://github.com/Nospoko/midi-translation/pull/11/files#diff-49c6dbffbc75a2442f1ed714ba8df982037d77248d9020291090508c21a792d7R33

If the lines
```python
    for path in shard_paths:
        for file in glob.glob(path):
        for file in glob.glob(f"{path}/*"):
            os.remove(file)
        os.rmdir(path)
```
were added after concatenation in shard_and_build, this code should produce error:
```python
midi_dataset = load_dataset(dataset_name, split=split)

num_shards = len(midi_dataset) // 5000 + 1 
translation_dataset = shard_and_build(
        dataset=midi_dataset,
        dataset_name=dataset_name,
        dataset_cfg=dataset_cfg,
        num_shards=num_shards,
        dataset_cache_path=dataset_cache_path,
)

train_translation_dataset, val_translation_dataset = translation_dataset.train_test_split(0.1).values()

```
It complained about not being able to find a directory of one of the shards - with "-part-0" suffix.

WojciechMat committed:
fix removing files, dataset_name inside dataset.info.splits['train']

WojciechMat commented:
The problem from the last comment was caused because of the `cache_files` attribute of a Dataset class. When calling `load_from_disc`, it remembered its path of origin and used it when spliting to train and test. 

We do want to keep the dataset in one folder, so the solution is to load the dataset from disc again right after saving:
```python
translation_dataset.save_to_disk(dataset_cache_path)
# load dataset again to update cache_files attribute
translation_dataset = Dataset.load_from_disk(dataset_cache_path)
```
another approach would probably be to update cache files mannually:
```python
cache_files = {}
for file in glob.glob(dataset_cache_path/data*of*)
    cache_files |= {"filename": file}
dataset.cache_files = cache_files
```

WojciechMat committed:
Merge branch 'MIDI-99/increase-dataset-size' of github.com:Nospoko/midi-translation into MIDI-99/increase-dataset-size

WojciechMat committed:
split-agnostic datasets

WojciechMat commented:
I've deleted option to not save te dataset, dataset_name is now stored in dataset.info.description to not mess with built-in hf functions such as dataset.shard, dataset is loaded again. You do not need to pass split argument, loading a dataset now looks like this:
```python
 translation_dataset = load_cache_dataset(
        dataset_cfg=cfg.dataset,
        dataset_name=cfg.dataset_name,
    )
    train_translation_dataset, val_translation_dataset = translation_dataset.train_test_split(0.1).values()

    if cfg.augmentation_rep > 0:
        train_translation_dataset = augment_dataset(
            dataset=train_translation_dataset,
            dataset_cfg=cfg.dataset,
            augmentation_probability=cfg.augmentation_probability,
            augmentation_rep=cfg.augmentation_rep,
        )

    train_dataset = MyTokenizedMidiDataset(
        dataset=train_translation_dataset,
        dataset_cfg=cfg.dataset,
        src_encoder=src_encoder,
        tgt_encoder=tgt_encoder,
    )
    val_dataset = MyTokenizedMidiDataset(
        dataset=val_translation_dataset,
        dataset_cfg=cfg.dataset,
        src_encoder=src_encoder,
        tgt_encoder=tgt_encoder,
    )
```
It is because maestro has validation and test splits, while giant datasets have only train split.

roszcz commented on data/dataset.py:
```diff
@@ -1,12 +1,15 @@
 import json
 import hashlib
 
+import datasets
```
```
data/dataset.py|6 col 1 error| 'datasets' imported but unused [F401]
```

keep your pre-commit hooks hooked :)

roszcz reviewed (commented):
The augmentation looks good, but as the main goal of this task, we should go towards training on a massive dataset, combining maestro, giant-midi, and pianofor.

In that case I think we should use maestro validation and test to test and validate, and use 100% of giant-midi and pianofor for training. In terms of musical content, maestro has much higher quality of data, so it makes sense to focus on it for validation.

WojciechMat committed:
training on mutliple datasets at once

WojciechMat committed:
augmentation on all cpus

WojciechMat commented:
1. I have implemented augmenting the dataset with all cpus using `dataset.map` !
2. test+validation of maestro-v1-sustain is now always used for validation.
3. augmentation is now specified in augmentation field of the config
4. you can concatenate datasets by using `dataset_name=my_dataset_path+my_other_dataset_path`
E.g. :
```shell
python train.py dataset_name=roszcz/maestro-v1-sustain+roszcz/giant-midi-sustain augmentation.repetitions=2 augmentation.shift=3 augmentation.probability=0.3 dataset.quantization.velocity=1
```
![image](https://github.com/Nospoko/midi-translation/assets/74838859/70129b34-2469-4767-b15c-5c2b76c6f21f)

WojciechMat committed:
Merge branch 'MIDI-99/increase-dataset-size' of github.com:Nospoko/midi-translation into MIDI-99/increase-dataset-size

WojciechMat committed:
warmup -> train.warmup, DRY

WojciechMat committed:
fix bug warmup -> train.warmup

WojciechMat commented:
I think we can merge it and start working on gradio app in another pull request. Or perhabs a new repository?

WojciechMat committed:
small fixes and updates
